django/db/models/query.py

```diff
@@ -608,7 +608,15 @@ def update_or_create(self, defaults=None, **kwargs):
                 return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
-            obj.save(using=self.db)
+            # `update_fields` does not support non-concrete fields yet (#31382)
+            fnames = set()
+            for f in self.model._meta.local_concrete_fields:
```

Comment:
I think this should include non-concrete local field as well since `select_for_update` will lock at tables involved in MTI.